src/shogun/labels/BinaryLabels.cpp

```diff
@@ -148,3 +149,55 @@ CLabels* CBinaryLabels::shallow_subset_copy()
 
 	return shallow_copy_labels;
 }
+
+namespace shogun
+{
+	Some<CBinaryLabels> binary_from_binary(CBinaryLabels* orig)
+	{
+		return Some<CBinaryLabels>::from_raw(orig);
+	}
+
+	Some<CBinaryLabels> binary_from_dense(CDenseLabels* orig)
+	{
+		auto labels = orig->get_labels();
+		for (auto i : range(labels.vlen))
+		{
+			if (!CMath::fequals(labels[i], 1.0, 0.0) &&
+			    !CMath::fequals(labels[i], -1.0, 0.0))
+			{
+				SG_SERROR(
+				    "Label[%d] is %f, but must be either +1.0 or -1.0.\n", i,
+				    labels[i])
+			}
+		}
+		auto result = new CBinaryLabels();
```

Comment:
wouldn't be better to actually drop those copy ctors that actually copy and fix it instead of this hack?

Reply:
Maybe, but I was afraid that this will lead to all sorts of other required changes in places that depend on the fact that labels copy the data.
I can look into it, though I think the patch is already big enough

Reply:
imo this is just nurturing the bad design so would be just better to drop that faulty copy-ctor.

Reply:
I agree, and will do it